### PR TITLE
CLN: dont consolidate in reshape.concat

### PR DESCRIPTION
**Pull request by jbrockmendel**
Looks like this consolidation was added here https://github.com/pandas-dev/pandas/commit/3b1c5b74e0ca5a782d4c070aac002b3e6d7e5290 in 2012, no clear reason why it is needed.  About to start an asv run.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
If you want to run benchmarks, I think you will need to run a dedicated benchmark where you create some non-consolidated data, as an impacted case is not necessarily covered by ASV.

**Comment by jreback** [comment]
> If you want to run benchmarks, I think you will need to run a dedicated benchmark where you create some non-consolidated data, as an impacted case is not necessarily covered by ASV.

if we don’t have asvs or tests then there is nothing to do and the change is fine. If you want to create and push an asv great but we cannot benchmark to some hypothetical asv.

**Comment by jorisvandenbossche** [comment]
> if we don’t have asvs or tests then there is nothing to do and the change is fine.

Sorry, but I completely disagree with that statement. Regularly on PRs that might impact performance, we ask PR authors to ensure there is a benchmark for the case they are changing. If there isn't yet one, we add one. 

For a general check, we can of course only rely on existing benchmarks, but for a PR that targets a very specific use case, we can test that specifically or add a benchmark.

**Comment by jreback** [comment]
my point which you missed that we already have tons of asvs for functions which hit this

but you don’t want to merge because we might impact some hypothetical which we can’t even right down?

**Comment by jorisvandenbossche** [comment]
I am not sure we have any asv's specifically for this (it might be, but can you then point to one?). 

I don't follow which hypothetical case you are talking about. You mean a non-consolidated dataframe? That's not hypothetical (eg add a column to a DataFrame, and you have a not-yet-consolidated dataframe). Why would we otherwise have those `consolidate` calls, if that would never have any effect?

Maybe in my original comment the "non-consolidated data" was a bit confusing. I am not talking about a hypothetical non-consolidating BlockManager, but about an actual, current DataFrame that is not consolidated at the moment when calling `concat`.

**Comment by TomAugspurger** [comment]
Something like

```python
In [21]: df = pd.DataFrame(index=list(range(100)))

In [22]: df1 = pd.DataFrame(index=list(range(100)))

In [23]: df2 = pd.DataFrame(index=list(range(100)))

In [24]: for i in range(10):
    ...:     df1[i] = np.random.randn(len(df))
    ...:     df2[i] = np.random.randn(len(df))
    ...:

In [25]: pd.concat([df1, df2])
```

Do we currently consolidate always in `concat`? Does it matter whether the non-concatenation axis is already aligned?

**Comment by jorisvandenbossche** [comment]
Indeed, something like that. 

BTW, with my original comment, I didn't mean to ask for much. Just a small example like the one Tom showed (but maybe a bit larger data), and do a `%timeit` on concat with master vs this branch. 
(which is actually much less than running the full ASV suite ..)

And quite probably, such a timing would show there isn't much impact (or maybe even faster without consolidating since it's a copy less?), but if we don't check that, we don't know. Such a small check is IMO a minimum for PRs like this.

**Comment by jbrockmendel** [comment]
hmm the 933x one makes zero sense, since it doesn't go through the affected code path.  Checking with timeit shows much smaller change, so writing that asv result off as incorrect

**Comment by jbrockmendel** [comment]
Using the example Tom suggested, this is appreciably slower:

```
df = pd.DataFrame(index=list(range(100)))

df1 = pd.DataFrame(index=list(range(100)))
df2 = pd.DataFrame(index=list(range(100)))

for i in range(10):
    df1[i] = np.random.randn(len(df))
    df2[i] = np.random.randn(len(df))


In [22]: %timeit pd.concat([df1, df2])
346 µs ± 3.55 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- master
1.34 ms ± 68.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR

[...] re-construct the non-consolidated DataFrames

In [23]: %timeit pd.concat([df1.copy(), df2.copy()])
830 µs ± 35.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- master
1.31 ms ± 17.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR
```

I expected the result in [22], am surprised by [23], will have to take a look at what is driving the results.

We need to decide how much (if any) perf penalty we're willing to accept to avoid side-effects.

**Comment by jbrockmendel** [comment]
Looks like the call to `concatenate_block_managers` is 8x slower non-consolidated, see below.

```
%prun -s cumulative for n in range(100): pd.concat([df1.copy(), df2.copy()])
```

PR:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.273    0.273 {built-in method builtins.exec}
        1    0.002    0.002    0.273    0.273 <string>:1(<module>)
      100    0.000    0.000    0.246    0.002 concat.py:67(concat)
      100    0.001    0.000    0.215    0.002 concat.py:451(get_result)
      100    0.004    0.000    0.200    0.002 concat.py:31(concatenate_block_managers)
     1000    0.002    0.000    0.072    0.000 concat.py:435(_is_uniform_join_units)
     3000    0.002    0.000    0.070    0.000 {built-in method builtins.all}
     3000    0.003    0.000    0.066    0.000 concat.py:450(<genexpr>)
     1100    0.005    0.000    0.064    0.000 concat.py:110(concat_compat)
     2000    0.008    0.000    0.063    0.000 concat.py:204(is_na)
     1100    0.006    0.000    0.043    0.000 concat.py:29(get_dtype_kinds)
     2000    0.001    0.000    0.041    0.000 missing.py:47(isna)
   126200    0.024    0.000    0.041    0.000 {built-in method builtins.isinstance}
     2000    0.003    0.000    0.041    0.000 missing.py:130(_isna)
     2000    0.006    0.000    0.034    0.000 missing.py:193(_isna_ndarraylike)
    15100    0.007    0.000    0.032    0.000 base.py:256(is_dtype)
      100    0.001    0.000    0.031    0.000 concat.py:292(__init__)
     3200    0.002    0.000    0.030    0.000 common.py:1180(needs_i8_conversion)
     1000    0.002    0.000    0.028    0.000 blocks.py:2705(make_block)
      100    0.000    0.000    0.028    0.000 concat.py:512(_get_new_axes)
```

master
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.161    0.161 {built-in method builtins.exec}
        1    0.001    0.001    0.161    0.161 <string>:1(<module>)
      100    0.000    0.000    0.139    0.001 concat.py:67(concat)
      100    0.002    0.000    0.098    0.001 concat.py:292(__init__)
      200    0.000    0.000    0.066    0.000 generic.py:5309(_consolidate)
      200    0.000    0.000    0.065    0.000 generic.py:5301(_consolidate_inplace)
      200    0.000    0.000    0.065    0.000 generic.py:5290(_protect_consolidate)
      200    0.000    0.000    0.065    0.000 generic.py:5304(f)
      200    0.000    0.000    0.063    0.000 managers.py:934(consolidate)
      200    0.000    0.000    0.056    0.000 managers.py:950(_consolidate_inplace)
      200    0.001    0.000    0.049    0.000 managers.py:1847(_consolidate)
      100    0.001    0.000    0.040    0.000 concat.py:453(get_result)
      100    0.000    0.000    0.029    0.000 concat.py:514(_get_new_axes)
      100    0.000    0.000    0.029    0.000 concat.py:517(<listcomp>)
     4000    0.001    0.000    0.027    0.000 managers.py:1852(<lambda>)
     4000    0.006    0.000    0.025    0.000 blocks.py:167(_consolidate_key)
      100    0.001    0.000    0.024    0.000 concat.py:31(concatenate_block_managers)
      300    0.001    0.000    0.023    0.000 base.py:4175(equals)
      200    0.001    0.000    0.021    0.000 generic.py:5652(copy)
      300    0.002    0.000    0.020    0.000 missing.py:358(array_equivalent)
      200    0.002    0.000    0.020    0.000 managers.py:1864(_merge_blocks)
      200    0.000    0.000    0.019    0.000 managers.py:752(copy)
     4000    0.006    0.000    0.018    0.000 _dtype.py:333(_name_get)
      200    0.002    0.000    0.016    0.000 managers.py:362(apply)
      100    0.000    0.000    0.016    0.000 concat.py:521(_get_comb_axis)
      100    0.000    0.000    0.016    0.000 api.py:65(get_objs_combined_axis)
      200    0.001    0.000    0.016    0.000 {built-in method builtins.sorted}
    44400    0.009    0.000    0.015    0.000 {built-in method builtins.isinstance}
      100    0.000    0.000    0.015    0.000 api.py:109(_get_combined_index)
     1400    0.001    0.000    0.014    0.000 common.py:1180(needs_i8_conversion)
      200    0.001    0.000    0.013    0.000 concat.py:110(concat_compat)
      100    0.000    0.000    0.013    0.000 concat.py:531(_get_concat_axis)
      100    0.000    0.000    0.012    0.000 concat.py:588(_concat_indexes)
      100    0.000    0.000    0.012    0.000 base.py:4112(append)
      300    0.001    0.000    0.012    0.000 blocks.py:2705(make_block)
     5100    0.003    0.000    0.012    0.000 base.py:256(is_dtype)
```

**Comment by jbrockmendel** [comment]
@jorisvandenbossche do you have a workaround in mind for how the all-1D case would avoid this perf hit?  if so, could that workaround be applicable here?

**Comment by TomAugspurger** [comment]
@jbrockmendel just to get a sense for order of magnitudes here: is the change in workloads roughly equivalent to the following:

```python
# consolidated: one (100, 10) array -> (200, 10) array
In [15]: a = np.ones((100, 10))

In [16]: %timeit np.concatenate([a, a])
1.86 µs ± 23.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

# non-consolidated: ten (100,) arrays -> ten (200,) arrays
In [17]: bs = [np.ones(100,) for _ in range(10)]

In [18]: %timeit [np.concatenate([b, b]) for b in bs]
15.9 µs ± 183 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

**Comment by jbrockmendel** [comment]
@TomAugspurger not sure which measurement your example is supposed to be analogous to.  The slowdown (8.5x) is pretty similar to the slowdown mentioned above inside `concatenate_block_managers`, but the actual user-facing slowdown is one of the smaller numbers (1.6x or 3.9x)

**Comment by jorisvandenbossche** [comment]
@TomAugspurger if you use slightly larger data, the python overhead (of the for loop + multiple function calls) decreases quickly:

```
In [10]: a = np.ones((10, 10000))  

In [11]: %timeit np.concatenate([a, a], axis=1)  
49.6 µs ± 1.68 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [12]: bs = [np.ones(10000) for _ in range(10)]  

In [13]: %timeit [np.concatenate([b, b]) for b in bs]  
55 µs ± 382 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Comment by TomAugspurger** [comment]
> not sure which measurement your example is supposed to be analogous to.

Just comparing the general idea of "concat this 2D array" vs. "concat these many 1D arrays" to get a sense for how things perform.

The raw number of rows and the relative number of rows to columns does matter to an extent.

concat

<details>

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from timeit import default_timer as tic

%matplotlib inline
import seaborn as sns
import matplotlib.pyplot as plt

n_rows = [10, 100, 1000, 10_000, 100_000]
n_cols = [10, 100, 1000]

timings = []

for n in n_rows:
    for c in n_cols:
        for i in range(3):
            a = np.ones((n, c))
            bs = [np.ones(n,) for _ in range(c)]
            t0 = tic()
            np.concatenate([a, a])
            t1 = tic()
            [np.concatenate([b, b]) for b in bs]
            t2 = tic()
            timings.append((n, c, i, 'consolidated', t1 - t0))
            timings.append((n, c, i, 'split', t2 - t1))


df = pd.DataFrame(timings, columns=['n_rows', 'n_cols', 'trial', 'policy', 'time'])
df

g = sns.FacetGrid(df, col="n_cols", hue="policy", )
g.map(
    sns.lineplot, 'n_rows', "time",
)
g.set(xscale="log", yscale="log")
g.add_legend()
```

</details>

**Comment by jbrockmendel** [comment]
So I definitely need to add an asv for this (based on https://github.com/pandas-dev/pandas/pull/34683#issuecomment-642343274).  Aside from that, we need to do one of three things:

1) Find a way to fix the performance hit in pd.concat (which I'm hoping @jorisvandenbossche has an idea for as I expect it will be the same for all-1D)
2) Decide we're OK with this performance hit.
3) Decide not to move forward with this.

**Comment by jbrockmendel** [comment]
@jorisvandenbossche in the ArrayManager PR you said that the code snippet here performed better than in master.  Any insight into improving the performance here?

**Comment by jbrockmendel** [comment]
Looks like in concatenate_block_managers it is going through concat_compat and making more copies

**Comment by jreback** [comment]
how's the asv's on this now?

**Comment by jbrockmendel** [comment]
This still looks about 25% slower than master on this benchmark https://github.com/pandas-dev/pandas/pull/34683#issuecomment-642343274
 
There's still some room to trim that down.

**Comment by jbrockmendel** [comment]
Just pushed, this is now slightly faster than master on the benchmark above https://github.com/pandas-dev/pandas/pull/34683#issuecomment-642343274

File: pandas/core/reshape/concat.py [hunk 1]
```diff
@@ -532,6 +531,7 @@ def _get_comb_axis(self, i: int) -> Index:
             copy=self.copy,
         )
 
+    @cache_readonly
     def _get_concat_axis(self) -> Index:
```
Comment (ivanovmg) [inline pandas/core/reshape/concat.py:535]: If it is property, then maybe ``_concat_axis``, without ``get`` would be better?

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -2669,35 +2676,37 @@ def get_block_type(values, dtype=None):
...
         cls = ObjectValuesExtensionBlock
-    elif is_extension_array_dtype(values.dtype):
+    elif isinstance(dtype, ExtensionDtype):
         cls = ExtensionBlock
-    elif issubclass(vtype, np.floating):
-        cls = FloatBlock
-    elif issubclass(vtype, np.timedelta64):
-        assert issubclass(vtype, np.integer)
+
+    elif kind == "M":
```
Comment (jreback) [inline pandas/core/internals/blocks.py:2699]: group with Timestamp?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:2699]: in principle third party EAs should go before this check (xref #34987)

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -2616,36 +2623,38 @@ def get_block_type(values, dtype=None):
...
     vtype = dtype.type
+    kind = dtype.kind
 
     cls: Type[Block]
 
     if is_sparse(dtype):
         # Need this first(ish) so that Sparse[datetime] is sparse
         cls = ExtensionBlock
-    elif is_categorical_dtype(values.dtype):
+    elif dtype.name == "category":
```
Comment (jreback) [inline pandas/core/internals/blocks.py:2637]: why not `is Categorical` ? e.g. since we are removing comparison vs 'category' generally (in your other PR)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:2637]: could do `isinstance(dtype, CategoricalDtype)`.  either way is fine by me

Comment (jreback) [reply pandas/core/internals/blocks.py:2637]: i think prefer that

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: updated+green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**